pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,282 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	attributeReferencerTagName = "attributereferencer"
+	errPanicedResolving        = "paniced while resolving references: %v"
+)
+
+// Error strings
+const (
+	errTaggedFieldlNotImplemented    = "ManagedReferenceResolver: the field has the %v tag, but has not implemented AttributeReferencer interface"
+	errBuildAttribute                = "ManagedReferenceResolver: could not build the attribute"
+	errAssignAttribute               = "ManagedReferenceResolver: could not assign the attribute"
+	errUpdateResourceAfterAssignment = "ManagedReferenceResolver: could not update the resource after resolving references"
+)
+
+// fieldHasTagPair is used in findAttributeReferencerFields
+type fieldHasTagPair struct {
+	fieldValue reflect.Value
+	hasTheTag  bool
+}
+
+// ReferenceStatusType is an enum type for the possible values for a Reference Status
+type ReferenceStatusType int
+
+const (
+	// ReferenceStatusUnknown is the default value
+	ReferenceStatusUnknown ReferenceStatusType = iota
+	// ReferenceNotFound shows that the reference is not found
+	ReferenceNotFound
+	// ReferenceNotReady shows that the reference is not ready
+	ReferenceNotReady
+	// ReferenceReady shows that the reference is ready
+	ReferenceReady
+)
+
+func (t ReferenceStatusType) String() string {
+	return []string{"Unknown", "NotFound", "NotReady", "Ready"}[t]
+}
+
+// ReferenceStatus has the name and status of a reference
+type ReferenceStatus struct {
+	Name   string
+	Status ReferenceStatusType
+}
+
+func (r ReferenceStatus) String() string {
+	return fmt.Sprintf("{reference:%s status:%s}", r.Name, r.Status)
+}
+
+// referencesAccessErr is used to indicate that one or more references can not
+// be accessed
+type referencesAccessErr struct {
+	statuses []ReferenceStatus
+}
+
+// newReferenceAccessErr returns a referencesAccessErr if any of the given
+// references are not ready
+func newReferenceAccessErr(statuses []ReferenceStatus) error {
+	for _, st := range statuses {
+		if st.Status != ReferenceReady {
+			return &referencesAccessErr{statuses}
+		}
+	}
+
+	return nil
+}
+
+func (r *referencesAccessErr) Error() string {
+	return fmt.Sprintf("%s", r.statuses)
+}
+
+// IsReferencesAccessError returns true if the given error is of type referencesAccessErr
+func IsReferencesAccessError(err error) bool {
+	_, result := err.(*referencesAccessErr)
+	return result
+}
+
+// CanReference is a type that is used as ReferenceResolver input
+type CanReference interface {
+	runtime.Object
+	metav1.Object
+}
```

Comment:
I mentioned this morning that I'd prefer we return to using `Managed` if we don't need to resolve references for resource classes. This was because I was under the mistaken impression that `CanReference` was a union of the `Managed` and `Class` interfaces. I see that it is in fact a much smaller interface, which I like. I think this interface makes sense even if we get rid of the class-level reference resolution.